### PR TITLE
Fixes and changes to resolution sequence for unresolved field and virtual calls on Z

### PR DESCRIPTION
**Pull request by r30shah**
While addressing the suggestions on fixing the GC map issue on Z in https://github.com/eclipse/openj9/pull/10599, compilation failure while building JDK on Z was introduced. Changes in this PR fixes that in addition switching to OMR::CCData API to allocate the codecache data for this resolution sequence.

Signed-off-by: Rahil Shah rahil@ca.ibm.com

### CODE REVIEW

**Comment by r30shah** [comment]
@fjeremic  Can I please get review on codegen changes?
@ymanton  Also can I get a review on the usage of OMR::CCData API ?

File: runtime/compiler/z/codegen/J9UnresolvedDataReadOnlySnippet.cpp [hunk 1]
```diff
@@ -105,8 +105,9 @@ J9::Z::UnresolvedDataReadOnlySnippet::emitSnippetBody()
 
    // Relative address to the start of the mainline resolution
    intptr_t startResolutionSeqLabelAddr = reinterpret_cast<intptr_t>(startResolveSequenceLabel->getCodeLocation());
-   TR_ASSERT_FATAL(cg()->canUseRelativeLongInstructions(startResolutionSeqLabelAddr), "startResolveSequenceLabel [%p] is outside relative immediate range", startResolutionSeqLabelAddr);*reinterpret_cast<int32_t*>(cursor) = static_cast<int32_t>(startOfResolutionSeqLabelAddr - (intptr_t)helperCallRA);
+   TR_ASSERT_FATAL(cg()->canUseRelativeLongInstructions(startResolutionSeqLabelAddr), "startResolveSequenceLabel [%p] is outside relative immediate range", startResolutionSeqLabelAddr);*reinterpret_cast<int32_t*>(cursor) = static_cast<int32_t>(startResolutionSeqLabelAddr - (intptr_t)helperCallRA);
```
Comment (fjeremic) [inline runtime/compiler/z/codegen/J9UnresolvedDataReadOnlySnippet.cpp:108]: This looks like two lines in one. Can we break them apart?

File: runtime/compiler/z/codegen/J9MemoryReference.cpp [hunk 2]
```diff
@@ -393,8 +397,11 @@ J9::Z::MemoryReference::create(TR::CodeGenerator* cg, TR::Node* node)
 
       typedef J9::Z::UnresolvedDataReadOnlySnippet::CCUnresolvedData CCUnresolvedData;
 
-      CCUnresolvedData* ccUnresolvedDataAddress =
-         reinterpret_cast<J9::Z::UnresolvedDataReadOnlySnippet::CCUnresolvedData*>(cg->allocateCodeMemory(sizeof(CCUnresolvedData), false));
+      OMR::CCData *codeCacheData = cg->getCodeCache()->manager()->getCodeCacheData();
+      OMR::CCData::index_t index;
+      codeCacheData->put(NULL, sizeof(CCUnresolvedData), 8, NULL, index);
+      
+      CCUnresolvedData* ccUnresolvedDataAddress = codeCacheData->get<CCUnresolvedData>(index);
```
Comment (ymanton) [inline runtime/compiler/z/codegen/J9MemoryReference.cpp:404]: You have to check the result of `put()` for failure, not `get()`.

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp [hunk 3]
```diff
@@ -1583,7 +1583,11 @@ J9::Z::PrivateLinkage::buildNoPatchingVirtualDispatchWithResolve(TR::Node *callN
       int32_t vtableOffset;
       };
    
-   ccResolveVirtualData *ccResolveVirtualDataAddress = reinterpret_cast<ccResolveVirtualData *>(cg()->allocateCodeMemory(sizeof(ccResolveVirtualData), false));
+   OMR::CCData *codeCacheData = cg()->getCodeCache()->manager()->getCodeCacheData();
+   OMR::CCData::index_t index;
+   codeCacheData->put(NULL, sizeof(ccResolveVirtualData), 8, NULL, index);
+   
+   ccResolveVirtualData *ccResolveVirtualDataAddress = codeCacheData->get<ccResolveVirtualData>(index);
```
Comment (ymanton) [inline runtime/compiler/z/codegen/S390PrivateLinkage.cpp:1590]: Likewise here.

**Review by ymanton — CHANGES_REQUESTED** [review]

File: runtime/compiler/z/codegen/J9UnresolvedDataReadOnlySnippet.cpp (hunk 1 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/J9UnresolvedDataReadOnlySnippet.cpp]: My apologies,  fixed in https://github.com/eclipse/openj9/compare/13a51f6bc611de8e95dcc43b18239f056fdbc616..e90116d7332681724cdd8aa492ff4522dadd0d6a

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp (hunk 3 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/S390PrivateLinkage.cpp]: Fixed in https://github.com/eclipse/openj9/compare/e90116d7332681724cdd8aa492ff4522dadd0d6a..69c427858ff45f769e7f9831d4e4777159a0a758

File: runtime/compiler/z/codegen/J9MemoryReference.cpp (hunk 2 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/J9MemoryReference.cpp]: Fixed in https://github.com/eclipse/openj9/compare/e90116d7332681724cdd8aa492ff4522dadd0d6a..69c427858ff45f769e7f9831d4e4777159a0a758

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp (hunk 3 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/S390PrivateLinkage.cpp]: Checking result of `put` in https://github.com/eclipse/openj9/compare/69c427858ff45f769e7f9831d4e4777159a0a758..e6572d23ab0cf110a98cc239bea656e8aa8729ec

File: runtime/compiler/z/codegen/J9MemoryReference.cpp (hunk 2 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/J9MemoryReference.cpp]: Checking result of `put` in https://github.com/eclipse/openj9/compare/69c427858ff45f769e7f9831d4e4777159a0a758..e6572d23ab0cf110a98cc239bea656e8aa8729ec

**Review by fjeremic — APPROVED** [review]

**Comment by r30shah** [comment]
@ymanton  Ping to check if you are OK with the changes?

File: runtime/compiler/z/codegen/J9MemoryReference.cpp [hunk 4]
```diff
@@ -393,14 +397,15 @@ J9::Z::MemoryReference::create(TR::CodeGenerator* cg, TR::Node* node)
...
-      CCUnresolvedData* ccUnresolvedDataAddress =
-         reinterpret_cast<J9::Z::UnresolvedDataReadOnlySnippet::CCUnresolvedData*>(cg->allocateCodeMemory(sizeof(CCUnresolvedData), false));
-
-      if (ccUnresolvedDataAddress == NULL)
+      OMR::CCData *codeCacheData = cg->getCodeCache()->manager()->getCodeCacheData();
+      OMR::CCData::index_t index;
+      if (!(codeCacheData->put(NULL, sizeof(CCUnresolvedData), 8, NULL, index)))
          {
-         comp->failCompilation<TR::CompilationException>("Could not allocate unresolved data metadata");
+         comp->failCompilation<TR::CompilationException>("Could not allocate uVdnresolved data metadata");
```
Comment (ymanton) [inline runtime/compiler/z/codegen/J9MemoryReference.cpp:404]: New typo: uVdnresolved

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp [hunk 5]
```diff
@@ -1583,12 +1583,14 @@ J9::Z::PrivateLinkage::buildNoPatchingVirtualDispatchWithResolve(TR::Node *callN
...
       };
    
-   ccResolveVirtualData *ccResolveVirtualDataAddress = reinterpret_cast<ccResolveVirtualData *>(cg()->allocateCodeMemory(sizeof(ccResolveVirtualData), false));
-
-   if (!ccResolveVirtualDataAddress)
+   OMR::CCData *codeCacheData = cg()->getCodeCache()->manager()->getCodeCacheData();
+   OMR::CCData::index_t index;
+   if (!(codeCacheData->put(NULL, sizeof(ccResolveVirtualData), 8, NULL, index)))
       {
       comp()->failCompilation<TR::CompilationException>("Could not allocate resovle virtual dispatch metadata");
```
Comment (ymanton) [inline runtime/compiler/z/codegen/S390PrivateLinkage.cpp:1590]: Existing typo: resovle

**Review by ymanton — CHANGES_REQUESTED** [review]
Thanks. It looks correct now. Made some minor suggestions in-line.

File: runtime/compiler/z/codegen/J9MemoryReference.cpp (hunk 4 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/J9MemoryReference.cpp]: fixed in https://github.com/eclipse/openj9/compare/e6572d23ab0cf110a98cc239bea656e8aa8729ec..6d1fee3c7c3d3066a711500dc5c93851f9df074c

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp (hunk 5 above)
Comment (r30shah) [reply runtime/compiler/z/codegen/S390PrivateLinkage.cpp]: Fixed in https://github.com/eclipse/openj9/compare/e6572d23ab0cf110a98cc239bea656e8aa8729ec..6d1fee3c7c3d3066a711500dc5c93851f9df074c

File: runtime/compiler/z/codegen/J9MemoryReference.cpp [hunk 6]
```diff
@@ -393,14 +397,15 @@ J9::Z::MemoryReference::create(TR::CodeGenerator* cg, TR::Node* node)
 
       typedef J9::Z::UnresolvedDataReadOnlySnippet::CCUnresolvedData CCUnresolvedData;
 
-      CCUnresolvedData* ccUnresolvedDataAddress =
-         reinterpret_cast<J9::Z::UnresolvedDataReadOnlySnippet::CCUnresolvedData*>(cg->allocateCodeMemory(sizeof(CCUnresolvedData), false));
-
-      if (ccUnresolvedDataAddress == NULL)
+      OMR::CCData *codeCacheData = cg->getCodeCache()->manager()->getCodeCacheData();
+      OMR::CCData::index_t index;
+      if (!(codeCacheData->put(NULL, sizeof(CCUnresolvedData), 8, NULL, index)))
```
[the remark this answers is not in the document]
Comment (r30shah) [reply runtime/compiler/z/codegen/J9MemoryReference.cpp]: Changed the alignment to alignof(CCUnresolvedData) in ccResolveVirtualData

File: runtime/compiler/z/codegen/S390PrivateLinkage.cpp [hunk 7]
```diff
@@ -1583,12 +1583,14 @@ J9::Z::PrivateLinkage::buildNoPatchingVirtualDispatchWithResolve(TR::Node *callN
       int32_t vtableOffset;
       };
    
-   ccResolveVirtualData *ccResolveVirtualDataAddress = reinterpret_cast<ccResolveVirtualData *>(cg()->allocateCodeMemory(sizeof(ccResolveVirtualData), false));
-
-   if (!ccResolveVirtualDataAddress)
+   OMR::CCData *codeCacheData = cg()->getCodeCache()->manager()->getCodeCacheData();
+   OMR::CCData::index_t index;
+   if (!(codeCacheData->put(NULL, sizeof(ccResolveVirtualData), 8, NULL, index)))
```
[the remark this answers is not in the document]
Comment (r30shah) [reply runtime/compiler/z/codegen/S390PrivateLinkage.cpp]: Fixed in https://github.com/eclipse/openj9/compare/69c427858ff45f769e7f9831d4e4777159a0a758..e6572d23ab0cf110a98cc239bea656e8aa8729ec

**Review by ymanton — APPROVED** [review]

**Pull request merged**